### ISSUE TITLE
Cannot use doc-view-mode properly

### ISSUE THREAD

**Issue by perryprog** [syl20bnr/spacemacs#7847]
#### Description :octocat:
When using C-c C-c to switch to viewing document in doc-view-mode, I get the error `Wrong type argument: stringp, nil`

#### Reproduction guide :beetle:
- Start Emacs
- Open any PostScript or PDF file
- Make sure you are in doc-view-mode
- Press `C-c C-c`

*Observed behaviour:* :eyes: :broken_heart:
There is an error in the minibuffer: `Wrong type argument: stringp, nil`

*Expected behaviour:* :heart: :smile:
I get a wonderfully formatted document with images and all.

#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.5
- Spacemacs branch: master (rev. 664ba6a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: hybrid
- Completion: helm
- Layers:
```elisp
(html helm auto-completion emacs-lisp git java clojure)
```
- System configuration features: JPEG RSVG NOTIFY ACL LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS

#### Backtrace :paw_prints:
```
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  insert-file-contents(nil nil nil nil t)
  #[0 "\304=\203\305 \210\306\307 \210\310\311!\207\312=\2030\313\306\314\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>\317\320!\203>\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/usr/local/Cellar/emacs/25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil]()
  ad-Advice-doc-view-toggle-display(#[0 "\304=\203\305 \210\306\307 \210\310\311!\207\312=\2030\313\306\314\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>\317\320!\203>\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/usr/local/Cellar/emacs/25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil])
  apply(ad-Advice-doc-view-toggle-display #[0 "\304=\203\305 \210\306\307 \210\310\311!\207\312=\2030\313\306\314\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>\317\320!\203>\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/usr/local/Cellar/emacs/25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil] nil)
  doc-view-toggle-display()
  funcall-interactively(doc-view-toggle-display)
  call-interactively(doc-view-toggle-display nil nil)
  command-execute(doc-view-toggle-display)
```

**Comment by bmag** [comment]
That's not a backtrace: https://github.com/syl20bnr/spacemacs/wiki/Debugging#get-the-backtrace

**Comment by perryprog** [comment]
My bad

**Comment by Geemili** [comment]
I get the same error when opening a pdf file.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.1
- Spacemacs: 0.200.5
- Spacemacs branch: master (rev. 664ba6a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(markdown helm emacs-lisp)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND DBUS GSETTINGS NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11

### Backtrace
```
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  insert-file-contents(nil nil nil nil t)
  #[0 \304=\203^R^@\305 \210\306^Q\307 \210\310\311!\20\312=\2030^@\313^Z\306^Q\314^K\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>^@\317\320!\203>^@\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/gnu/store/5rzrpgjag3h4ybf1s6rg19m2q1jjkr9w-emacs-25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil]()
  ad-Advice-doc-view-toggle-display(#[0 \304=\203^R^@\305 \210\306^Q\307 \210\310\311!\20\312=\2030^@\313^Z\306^Q\314^K\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>^@\317\320!\203>^@\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/gnu/store/5rzrpgjag3h4ybf1s6rg19m2q1jjkr9w-emacs-25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil])
  apply(ad-Advice-doc-view-toggle-display #[0 \304=\203^R^@\305 \210\306^Q\307 \210\310\311!\20\312=\2030^@\313^Z\306^Q\314^K\306\211\211\313%\210\307 \210\310\311!\210\315\306!)\207\316 \203>^@\317\320!\203>^@\321 \210\304 \207" [major-mode buffer-read-only buffer-undo-list doc-view--buffer-file-name doc-view-mode doc-view-kill-proc nil doc-view-fallback-mode doc-view-minor-mode 1 text-mode t insert-file-contents set-buffer-modified-p buffer-modified-p y-or-n-p "The buffer has been modified.  Save the changes? " save-buffer] 6 ("/gnu/store/5rzrpgjag3h4ybf1s6rg19m2q1jjkr9w-emacs-25.1/share/emacs/25.1/lisp/doc-view.elc" . 41583) nil] nil)
  doc-view-toggle-display()
  funcall-interactively(doc-view-toggle-display)
  call-interactively(doc-view-toggle-display nil nil)
  command-execute(doc-view-toggle-display)
```

### Side Note
I had trouble copying the backtrace. It only copied up to a certain point, and then stopped. I got around this by saving the backtrace into a file, and then using less to view it. I then copied the text onto the clipboard using `gnome-terminal`'s copy function.

**Comment by perryprog** [comment]
Hmmm... good to know. I have had this problem with Aquamacs and vanilla Emacs before so this might be a problem with `doc-view-mode`.

**Comment by Geemili** [comment]
**tl;dr** You may need to install GhostScript.

Oh, I figured it out. At least for myself.

I am using [GuixSD](https://www.gnu.org/software/guix/), and the emacs package doesn't include optional packages. When I searched up "opening pdf in emacs", I found [this Stack Overflow answer](http://stackoverflow.com/questions/28474736/how-to-open-pdf-file-in-emacs), which led to the [emacs manual about doc-view](https://www.gnu.org/software/emacs/manual/html_node/emacs/Document-View.html):

> ### Footnotes
> ### (16)
>
> For PostScript files, GhostScript is a hard requirement. For DVI files, dvipdf or dvipdfm is needed. For OpenDocument and Microsoft Office documents, the unoconv tool is needed.

I looked at the `emacs` package for Guix; it does not include GhostScript. Installing GhostScript solved the problem.

**Comment by perryprog** [comment]
__cries__

Besides, I'm trying to open PDFs. So strange...

**Comment by Geemili** [comment]
Okay, just tried it again, this time opening the file from `Dired`. It opens into `doc-view-mode` perfectly. Pressing `C-c C-c` switches into viewing the pdf's source.

Trying to get back into `doc-view-mode` doesn't work.

So,
1. Open emacs
2. Type `M-x dired`
3. Navigate to PDF
4. Open it. It should open in `dov-view` mode.
5. Press `C-c C-c`
6. Press `C-c C-c`
7. Error!

**Comment by perryprog** [comment]
Nice work. I recently got a new MacBook, and I have not set up emacs on it
yet so I can't check this.

**Comment by perryprog** [comment]
Nice work. I recently got a new MacBook, and I have not set up Emacs yet, But I will try to set it up later and see if I can recreate that.

**Comment by LdBeth** [comment]
Guys, remember to install the deps every time updated `pdf-tool`, the full instruction is available here: <https://github.com/politza/pdf-tools#server-prerequisites>

Also the layer's doc has some instructions about this.

**Comment by bleggett** [comment]
Still am seeing @Geemili's error. PDF is shown initially, can't toggle to text and back without the `stringp, nil` error.

**Comment by Miciah** [comment]
The issue appears to be resolved in Emac 27 with http://git.savannah.gnu.org/cgit/emacs.git/commit/?id=affd2a0cee901404251ea1d2ef01bfb618f5213a.

**Comment by perryprog** [comment]
Sorry for the late reply. I can confirm @Miciah's solution fixes this for me. Thanks a bunch!

Going to leave this open as this is just a workaround of course, unless someone thinks otherwise.
